test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+// Copyright IBM Corp. 2013,2016. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+var assert = require('assert');
+var expect = require('./helpers/expect');
+var request = require('supertest');
+var loopback = require('../');
+var User, User2, AccessToken;
+var async = require('async');
+
+describe('User', function() {
+  this.timeout(10000);
+
+  var validCredentialsEmail = 'foo@bar.com';
+  var validCredentials = {email: validCredentialsEmail, password: 'bar'};
+  var validCredentialsWithRealm = {
+    email: 'foo3@bar.com', password: 'bar', realm: 'foobar'};
+  var validCredentialsWithTTL = {email: 'foo@bar.com', password: 'bar', ttl: 3600};
+  var validCredentialsWithTTLAndScope = {
+    email: 'foo@bar.com', password: 'bar', ttl: 3600, scope: 'all'};
+  var validMixedCaseEmailCredentials = {email: 'Foo@bar.com', password: 'bar'};
+  var invalidCredentials = {email: 'foo1@bar.com', password: 'invalid'};
+  var incompleteCredentials = {password: 'bar1'};
+
+  var validCredentialsUser = null;
+  var validCredentialsUser2 = null;
+
+  // Create a local app variable to prevent clashes with the global
+  // variable shared by all tests. While this should not be necessary if
+  // the tests were written correctly, it turns out that's not the case :(
+  var app;
+
+  beforeEach(function setupAppAndModels(done) {
+    // override the global app object provided by test/support.js
+    // and create a local one that does not share state with other tests
+    app = loopback({localRegistry: true, loadBuiltinModels: true});
+    app.dataSource('db', {connector: 'memory'});
+
+    // attach User and related models
+    User = app.registry.createModel({
+      name: 'TestUser',
+      base: 'User',
+      // forceId is set to false for the purpose of updating the same affected user within the
+      // `Email Update` test cases.
+      forceId: false,
+      // Speed up the password hashing algorithm for tests
+      saltWorkFactor: 4,
+    });
+    app.model(User, {dataSource: 'db'});
+
+    // create and attach a second User-based model
+    User2 = app.registry.createModel({
```

Comment:
I would prefer more descriptive names if possible. For example `Customer` and `Employee`? Or at least `AnotherUser` instead of `User2`. Also please change model names to match the variable name (or vice versa) - e.g. `TestUser` + `AnotherUser`.